PasswordRecipientInfoBuilder for CMS

bkstein opened the pull request:
This provides a builder for CMS' `PasswordRecipientInfo` according to RFC 3211.
In contrast to the `KeyTransRecipientInfoBuilder` the key-encryption algorithm must be provided by the user (who has to provide an implementation of `PwriEncryptor`). This allows for more flexibility in choosing allgorithms for key derivation and encryption.
An example for using the implementation is in the new test.

baloo commented on cms/src/builder.rs:
```diff
@@ -705,32 +704,106 @@ impl RecipientInfoBuilder for KekRecipientInfoBuilder {
         })
     }
+
+    /// Wrap the content-encryption key according to RFC 3211, §2.3.1:
+    ///     ....
+    ///     The formatted CEK block then looks as follows:
+    ///     CEK byte count || check value || CEK || padding (if required)
```
```suggestion
    /// Wrap the content-encryption key according to [RFC 3211, §2.3.1]:
    ///     ....
    ///     The formatted CEK block then looks as follows:
    ///     CEK byte count || check value || CEK || padding (if required)
    ///
    /// [RFC 3211, §2.3.1]: https://www.rfc-editor.org/rfc/rfc3211#section-2.3.1
```

baloo commented on cms/src/builder.rs:
```diff
@@ -705,32 +704,106 @@ impl RecipientInfoBuilder for KekRecipientInfoBuilder {
+        let cek_byte_count: u8 = content_encryption_key.len().try_into().map_err(|_| {
+            Error::Builder("Content encryption key length must not exceed 255".to_string())
+        })?;
+        let mut check_value = vec![
+            0xff ^ content_encryption_key[0],
+            0xff ^ content_encryption_key[1],
+            0xff ^ content_encryption_key[2],
+        ];
+
+        let mut wrapped_cek: Vec<u8> = Vec::new();
+        wrapped_cek.insert(0, cek_byte_count);
+        wrapped_cek.append(&mut check_value);
```
```suggestion
        let mut wrapped_cek: Vec<u8> = Vec::with_capacity(4 + padding_length);
        wrapped_cek.insert(0, cek_byte_count);
        wrapped_cek.insert(0xff ^ content_encryption_key[0]);
        wrapped_cek.insert(0xff ^ content_encryption_key[1]);
        wrapped_cek.insert(0xff ^ content_encryption_key[2]);
```

baloo commented:
(@bkstein if you rebase, the security audit error should go away. This is not actionable for now and we'll ignore it until RSA 0.10 is released)

baloo commented on der/src/asn1/octet_string.rs:
```diff
@@ -244,13 +244,15 @@ mod bytes {
 #[allow(clippy::unwrap_used)]
 mod tests {
-    use crate::asn1::{OctetStringRef, PrintableStringRef};
+    use crate::asn1::{OctetString, OctetStringRef, PrintableStringRef};
```
I think this fails in tests because it would need to be gated behind `#[cfg(feature = "alloc")]`, but what were you trying to test?

bkstein replied:
I'm sorry. That was just for debugging OctetString and I forgot to mark it for removal before pushing. (Normally I use `// TODO NC` for that and I have a pre-commit git hook checking it.)

baloo commented on cms/src/builder.rs:
```diff
@@ -705,32 +704,106 @@ impl RecipientInfoBuilder for KekRecipientInfoBuilder {
+    /// Provided password encryptor
+    pub key_encryptor: P,
+    /// Random number generator
+    pub rng: &'r mut R,
```
the only solution I could come up with is this one: https://github.com/bkstein/RustCrypto-formats/pull/1
But this introduces some API breakage in the builder (but we can just push that in a cms-0.3)

baloo commented on cms/src/builder.rs:
```diff
@@ -705,32 +704,104 @@ impl RecipientInfoBuilder for KekRecipientInfoBuilder {
+        wrapped_cek.push(cek_byte_count);
+        wrapped_cek.push(0xff ^ content_encryption_key[0]);
+        wrapped_cek.push(0xff ^ content_encryption_key[1]);
+        wrapped_cek.push(0xff ^ content_encryption_key[2]);
+        if padding_length > 0 {
```
```suggestion
        wrapped_cek.push(0xff ^ content_encryption_key[2]);
        wrapped_cek.append(content_encryption_key);
        if padding_length > 0 {
```

We forgot the CEK here right?

bkstein replied:
That could explain, why decryption failed last night :smiley:

baloo replied:
hu, sorry I always get mixed up with the vec methods.

bkstein commented:
@baloo First, I apologize for the long time beeing absent. We were busy working on our project and we had this PR as a local patch. But I'm interested in making this official and would like to resume. I started by updating to the master branch and experienced problems with the new crate versions. The clippy error is due to incompatible versioning of some Dependencies (digest is referenced by ecdsa with version v0.11.0-pre.8, but I introduced pkdf2, which references digest v0.10.7):
 ```
$ cargo tree
...
│   │       [dev-dependencies]
│   │       ├── ecdsa v0.17.0-pre.5 (https://github.com/RustCrypto/signatures#c2f3ee64)
...
│   │       │   ├── digest v0.11.0-pre.8
...
│   │       │   ├── elliptic-curve v0.14.0-pre.5
...
│   │       │   │   ├── digest v0.11.0-pre.8 (*)
...
│   │       │   │   ├── pkcs8 v0.11.0-pre.0 (/home/bkstein/Projects/github.com/RustCrypto/formats/pkcs8)
...
│   │       │   │   │   ├── pkcs5 v0.8.0-pre.0 (/home/bkstein/Projects/github.com/RustCrypto/formats/pkcs5)
...
│   │       │   │   │   │   ├── pbkdf2 v0.12.2
│   │       │   │   │   │   │   ├── digest v0.10.7
...
```
At the moment, I don't see how to resolve this.

baloo commented:
Ha, this is an oversight. I've bumped a bunch of dependencies in the ecosystem to pre-release versions but forgot pkcs5.

See; https://github.com/RustCrypto/formats/pull/1391

bkstein commented:
@baloo I updated this PR to the latest master and dependencies. May be merged from my side.

bkstein commented:
@baloo I tried to write a test for the `PasswordRecipientInfoBuilder` and hit the problem, that my current implementation needs two separate rng's. I did not find a way around this problem and would like to resume your proposal of using `build_with_rng()` methods. Do you still have that branch? I know, you deleted it on github. If not, I still have it and could reincarnate it, but ownership would be mine, then. Were that ok for you?

baloo commented:
> Do you still have that branch? I know, you deleted it on github.

You can grab it back in anycase:
```
git fetch https://github.com/bkstein/RustCrypto-formats.git refs/pull/1/head:baloo/cms/pwri-builder
```

> I still have it and could reincarnate it, but ownership would be mine, then. Were that ok for you?
Yes! Please grab it!

bkstein commented:
@baloo I think, this PR is ready. However, there is a problem in CI with cargo-audit which is not related to my changes, I think.

tarcieri commented:
@bkstein rebasing should fix the audit

bkstein commented:
@baloo @tarcieri I almost forgot this PR. It is ready since last October. I implemented baloo's proposal (-> don't store a mutable reference to an rng in the builder struct) and rebased the branch. It's ready from my side (if the pipeline succeeds).
